password_security/models/res_users.py

```diff
@@ -0,0 +1,218 @@
+
+        estimation = self.get_estimation(password)
+        if estimation["score"] < company_id.password_estimate:
+            if estimation["feedback"]["warning"]:
+                raise UserError(estimation["feedback"]["warning"])
+            else:
+                raise UserError(_("Choose a stronger password!"))
+
+        return True
+
+    def _password_has_expired(self):
+        self.ensure_one()
+        if not self.password_write_date:
+            return True
+
+        if not self.company_id.password_expiration:
+            return False
+
+        days = (fields.Datetime.now() - self.password_write_date).days
+        return days > self.company_id.password_expiration
+
+    def action_expire_password(self):
+        expiration = delta_now(days=+1)
+        for user in self:
+            user.mapped("partner_id").signup_prepare(
+                signup_type="reset", expiration=expiration
+            )
+
+    def _validate_pass_reset(self):
+        """It provides validations before initiating a pass reset email
+        :raises: UserError on invalidated pass reset attempt
+        :return: True on allowed reset
+        """
+        for user in self:
+            pass_min = user.company_id.password_minimum
+            if pass_min <= 0:
+                continue
+            write_date = user.password_write_date
+            delta = timedelta(hours=pass_min)
+            if write_date + delta > datetime.now():
+                raise UserError(
+                    _(
+                        "Passwords can only be reset every %d hour(s). "
+                        "Please contact an administrator for assistance."
+                    )
+                    % pass_min
+                )
+        return True
+
+    def _check_password_history(self, password):
+        """It validates proposed password against existing history
+        :raises: UserError on reused password
+        """
+        crypt = self._crypt_context()
+        for user in self:
+            password_history = user.company_id.password_history
+            if not password_history:  # disabled
+                recent_passes = self.env["res.users.pass.history"]
+            elif password_history < 0:  # unlimited
+                recent_passes = user.password_history_ids
+            else:
+                recent_passes = user.password_history_ids[:password_history]
+            if recent_passes.filtered(
+                lambda r: crypt.verify(password, r.password_crypt)
+            ):
+                raise UserError(
+                    _("Cannot use the most recent %d passwords")
+                    % user.company_id.password_history
+                )
+
+    def _set_encrypted_password(self, uid, pw):
+        """It saves password crypt history for history rules"""
+        res = super(ResUsers, self)._set_encrypted_password(uid, pw)
+
+        self.write({"password_history_ids": [(0, 0, {"password_crypt": pw})]})
```

Comment:
This method can be called as a (portal) user from `_check_credentials` with a replacement hash. In that case, the user does not have permissions to write.

```suggestion
        self.sudo().write({"password_history_ids": [(0, 0, {"password_crypt": pw})]})
```

Reply:
@kmagusiak thank you!
I made a fix in commit https://github.com/OCA/server-auth/pull/482/commits/4b1b2686d841b73f2594103d8218a17c989fa0c5 by setting the proper rights for portal user. Is that working for you?

Reply:
I haven't tested, but it looks like it should work.